eth/tracers/api_blockResult.go

```diff
@@ -0,0 +1,354 @@
+		Nonce:    state.GetNonce(from),
+		Balance:  (*hexutil.Big)(state.GetBalance(from)),
+		CodeHash: state.GetCodeHash(from),
+	}
+	var receiver *types.AccountWrapper
+	if to != nil {
+		receiver = &types.AccountWrapper{
+			Address:  *to,
+			Nonce:    state.GetNonce(*to),
+			Balance:  (*hexutil.Big)(state.GetBalance(*to)),
+			CodeHash: state.GetCodeHash(*to),
+		}
+	}
+	// Run the transaction with tracing enabled.
+	vmenv := vm.NewEVM(env.blockCtx, core.NewEVMTxContext(msg), state, api.backend.ChainConfig(), vm.Config{Debug: true, Tracer: env.tracer, NoBaseFee: true})
+	// Computes the new state by applying the given message.
+	env.rMu.Lock()
+	result, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+	if err != nil {
+		return fmt.Errorf("tracing failed: %w", err)
+	}
+	env.rMu.Unlock()
+	// If the result contains a revert reason, return it.
+	returnVal := result.Return()
+	if len(result.Revert()) > 0 {
+		returnVal = result.Revert()
+	}
+
+	createdAcc := env.tracer.CreatedAccount()
+	var after []*types.AccountWrapper
+	if to == nil {
+		if createdAcc == nil {
+			return errors.New("unexpected tx: address for created contract unavailable")
+		}
+		to = &createdAcc.Address
+	}
+	// collect affected account after tx being applied
+	for _, acc := range []common.Address{from, *to, api.backend.Coinbase()} {
+		after = append(after, &types.AccountWrapper{
+			Address:  acc,
+			Nonce:    state.GetNonce(acc),
+			Balance:  (*hexutil.Big)(state.GetBalance(acc)),
+			CodeHash: state.GetCodeHash(acc),
+		})
+	}
+
+	// merge required proof data
+	proofAccounts := env.tracer.UpdatedAccounts()
+	for addr := range proofAccounts {
+		addrStr := addr.String()
+
+		env.pMu.Lock()
+		_, existed := env.Proofs[addrStr]
+		env.pMu.Unlock()
+		if existed {
+			continue
+		}
+		proof, err := state.GetProof(addr)
+		if err != nil {
+			log.Error("Proof not available", "address", addrStr, "error", err)
+			// but we still mark the proofs map with nil array
+		}
+		wrappedProof := make([]hexutil.Bytes, len(proof))
+		for i, bt := range proof {
+			wrappedProof[i] = bt
+		}
+		env.pMu.Lock()
+		env.Proofs[addrStr] = wrappedProof
+		env.pMu.Unlock()
+	}
+
+	proofStorages := env.tracer.UpdatedStorages()
+	for addr, keys := range proofStorages {
+		for key := range keys {
+			addrStr := addr.String()
+			env.sMu.Lock()
+			m, existed := env.StorageProofs[addrStr]
+			if !existed {
+				m = make(map[string][]hexutil.Bytes)
+				env.StorageProofs[addrStr] = m
+			}
+			env.sMu.Unlock()
+
+			keyStr := key.String()
+			if _, existed := m[keyStr]; !existed {
+				proof, err := state.GetStorageTrieProof(addr, key)
+				if err != nil {
+					log.Error("Storage proof not available", "error", err, "address", addrStr, "key", keyStr)
+					// but we still mark the proofs map with nil array
+				}
+				wrappedProof := make([]hexutil.Bytes, len(proof))
+				for i, bt := range proof {
+					wrappedProof[i] = bt
+				}
+				m[keyStr] = wrappedProof
```

Comment:
1. Why not lock here, what if two concurrent transactions both modify the same contract storage entry?
2. Same question as above, shouldn't we call `state.GetStorageTrieProof` on the base state?